File: HandleCellEvents.vb
Imports System.Windows.Forms
Imports System.Drawing
Imports System.Drawing.Graphics
Imports System.Threading
Class KeyPressAwareDataGridView
    Inherits DataGridView
    Protected Friend blnNav As Boolean = True
    Protected Overloads Overrides Sub OnControlAdded(ByVal e As ControlEventArgs)
        Me.subscribeEvents(e.Control)
        MyBase.OnControlAdded(e)
    End Sub

    Protected Overloads Overrides Sub OnControlRemoved(ByVal e As ControlEventArgs)
        Me.unsubscribeEvents(e.Control)
        MyBase.OnControlRemoved(e)
    End Sub

    Protected Overloads Overrides Function ProcessDataGridViewKey(ByVal e As KeyEventArgs) As Boolean
        Dim procesedInternally As Boolean = False

        'If keyDownHook IsNot Nothing Then
        RaiseEvent keyDownHook(Me, e)
        procesedInternally = e.SuppressKeyPress
        'End If
        If e.KeyCode = Keys.Enter Or e.KeyCode = Keys.Tab Then
            ' MsgBox(blnNav)
            ' If blnNav Then
            If Me.CurrentCell.ColumnIndex = 1 Then
                If Not Me.CurrentRow Is Nothing Then
                    Me.ProcessRightKey(e.KeyData)
                    Me.CurrentCell = Me.CurrentRow.Cells(5)
                    Me.ProcessF2Key(e.KeyData)
                    Return True
                End If
            ElseIf Me.CurrentCell.ColumnIndex = 5 Then
                If Val(Me.CurrentCell.EditedFormattedValue) = 0 Then Return MyBase.ProcessDataGridViewKey(e)
                Me.ProcessTabKey(e.KeyData)
                Me.ProcessTabKey(e.KeyData)
                Me.ProcessTabKey(e.KeyData)
                Me.ProcessF2Key(e.KeyData)
                Return True
            Else
                If e.KeyCode = Keys.Enter Then
                    Me.ProcessTabKey(e.KeyData)
                    Return True
                End If
                'If Me.CurrentCell.ColumnIndex <> 5 And Me.CurrentCell.ColumnIndex <> 6 Then
                '    Me.ProcessRightKey(e.KeyData)
                '    Return True
                'End If
            End If
            'Else
            '    blnNav = True
            '    Return True
            'End If
        End If


        If procesedInternally Then
            Return True
        Else
            Return MyBase.ProcessDataGridViewKey(e)
        End If
    End Function
    Protected Overrides Function ProcessDialogKey(ByVal keyData As Keys) As Boolean
        'Dim e As KeyEventArgs
        'RaiseEvent keyDownHook(Me, e.)
        ' Extract the key code from the key value. 
        Dim key As Keys = keyData And Keys.KeyCode
        ' Handle the ENTER key as if it were a RIGHT ARROW key. 

        If key = Keys.Enter Or key = Keys.Tab Then
            ' MsgBox(blnNav)
            'If blnNav Then
            If Me.CurrentCell.ColumnIndex = 1 Then
                If Not Me.CurrentRow Is Nothing Then
                    Me.ProcessRightKey(keyData)
                    Me.CurrentCell = Me.CurrentRow.Cells(5)
                    Me.ProcessF2Key(keyData)
                    Return True
                End If
            ElseIf Me.CurrentCell.ColumnIndex = 5 Then
                If Val(Me.CurrentCell.EditedFormattedValue) = 0 Then Return MyBase.ProcessDialogKey(keyData)
                Me.ProcessTabKey(keyData)
                Me.ProcessTabKey(keyData)
                Me.ProcessTabKey(keyData)
                ' MsgBox(key)
                Me.ProcessF2Key(keyData)
                Return True
            Else
                If key = Keys.Enter Then
                    Me.ProcessTabKey(keyData)
                    Return True
                End If

                'If Me.CurrentCell.ColumnIndex <> 5 And Me.CurrentCell.ColumnIndex <> 6 Then
                '    Me.ProcessRightKey(keyData)
                '    Return True
                'End If
            End If
            'Else
            '    blnNav = True
            '    Return True
            'End If
        End If

        Return MyBase.ProcessDialogKey(keyData)

    End Function

    Private Sub subscribeEvents(ByVal control As Control)
        AddHandler control.KeyDown, AddressOf Me.control_KeyDown
        AddHandler control.KeyPress, AddressOf Me.control_keyPress

        AddHandler control.ControlAdded, AddressOf Me.control_ControlAdded
        AddHandler control.ControlRemoved, AddressOf Me.control_ControlRemoved

        For Each innerControl As Control In control.Controls
            Me.subscribeEvents(innerControl)
        Next
    End Sub

    Private Sub unsubscribeEvents(ByVal control As Control)
        RemoveHandler control.KeyDown, AddressOf Me.control_KeyDown
        RemoveHandler control.KeyPress, AddressOf Me.control_keyPress

        RemoveHandler control.ControlAdded, AddressOf Me.control_ControlAdded
        RemoveHandler control.ControlRemoved, AddressOf Me.control_ControlRemoved

        For Each innerControl As Control In control.Controls
            Me.unsubscribeEvents(innerControl)
        Next
    End Sub

    Private Sub control_ControlAdded(ByVal sender As Object, ByVal e As ControlEventArgs)
        Me.subscribeEvents(e.Control)
    End Sub

    Private Sub control_ControlRemoved(ByVal sender As Object, ByVal e As ControlEventArgs)
        Me.unsubscribeEvents(e.Control)
    End Sub

    Private Sub control_KeyDown(ByVal sender As Object, ByVal e As KeyEventArgs)
        RaiseEvent keyDownHook(sender, e)
    End Sub
    Private Sub control_keyPress(ByVal sender As Object, ByVal e As KeyPressEventArgs)
        RaiseEvent keyPressHook(sender, e)
    End Sub

    Public Event keyPressHook As KeyPressEventHandler
    Public Event keyDownHook As KeyEventHandler

End Class

File: HandleCellEvents.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Windows.Forms;
using System.Drawing;
using System.Threading;

namespace BillingManager
{
    class KeyPressAwareDataGridView : DataGridView
    {

    protected Boolean blnNav = true;
    public event KeyEventHandler keyDownHook;  
    public event KeyPressEventHandler keyPressHook;
    protected override void OnControlAdded( ControlEventArgs e)
    {
        subscribeEvents(e.Control);
        base.OnControlAdded(e);
    }

    protected override void OnControlRemoved(ControlEventArgs e)
    {
        unsubscribeEvents(e.Control);
        base.OnControlRemoved(e);
    }

    protected override Boolean ProcessDataGridViewKey(KeyEventArgs e) 
    {
        Boolean procesedInternally = false;

        keyDownHook(this, e);
        procesedInternally = e.SuppressKeyPress;
        if (e.Handled)
        {
            if (procesedInternally)
                return true;
            else
                return base.ProcessDataGridViewKey(e);
        }
        if( e.KeyCode == Keys.Enter || e.KeyCode == Keys.Tab)
        {
            if(this.CurrentCell.ColumnIndex == 2)
            {
                if(this.CurrentRow!=null)
                {
                    ProcessRightKey(e.KeyData);
                    if (!this.IsCurrentCellDirty)
                    {
                        this.CurrentCell = this.CurrentRow.Cells[5];
                        ProcessF2Key(e.KeyData);
                    }
                    return true;
                }
            }
            else if (this.CurrentCell.ColumnIndex == 3)
            {
                if (this.CurrentCell.Value == null) { return false; }
                if (this.CurrentRow != null)
                {
                    ProcessRightKey(e.KeyData);
                    if (!this.IsCurrentCellDirty)
                    {
                        this.CurrentCell = this.CurrentRow.Cells[5];
                        ProcessF2Key(e.KeyData);
                    }
                    return true;
                }
            }
            else if( this.CurrentCell.ColumnIndex == 5)
            {
                double val=0;
                double.TryParse(string.Concat(CurrentCell.EditedFormattedValue), out val);
                if( val == 0)return base.ProcessDataGridViewKey(e);
                ProcessTabKey(e.KeyData);
                ProcessTabKey(e.KeyData);
                ProcessTabKey(e.KeyData);
                ProcessTabKey(e.KeyData);
                ProcessTabKey(e.KeyData);
                ProcessF2Key(e.KeyData);
                return true;
            }
            else
            {
                if( e.KeyCode == Keys.Enter)
                {
                    ProcessTabKey(e.KeyData);
                    return true;
                }
            }
        }


        if(procesedInternally)
            return true;
        else
            return base.ProcessDataGridViewKey(e);
        
    }

    protected override Boolean ProcessDialogKey(Keys keyData) 
    {
         Keys key = keyData & Keys.KeyCode;
        //' Handle the ENTER key as if it were a RIGHT ARROW key. 

        if(key == Keys.Enter || key == Keys.Tab)
        {
            if (CurrentCell.ColumnIndex == 2)
            {
                if(CurrentRow !=null)
                {
                    if (EditingControl.Text == "?")
                    {
                         return false;
                    }
                    ProcessRightKey(keyData);
                    if (!this.IsCurrentCellDirty)
                    {
                        CurrentCell = CurrentRow.Cells[3];
                        ProcessF2Key(keyData);
                        DataGridViewTextBoxEditingControl editControl = (DataGridViewTextBoxEditingControl)EditingControl;
                        if (editControl != null)
                        {
                            editControl.SelectionStart = 0;
                            editControl.SelectionLength = editControl.Text.Length;
                        }
                    }
                    return true;
                }
            }
            else if (CurrentCell.ColumnIndex == 3)
            {
                if (CurrentRow != null)
                {
                    if (EditingControl.Text == "?")
                    {
                        return false;
                    }
                    if (EditingControl.Text == "")
                    {
                        return false;
                    }
                    ProcessRightKey(keyData);
                    if (!this.IsCurrentCellDirty)
                    {
                        CurrentCell = CurrentRow.Cells[5];
                        ProcessF2Key(keyData);
                        DataGridViewTextBoxEditingControl editControl = (DataGridViewTextBoxEditingControl)EditingControl;
                        if (editControl != null)
                        {
                            editControl.SelectionStart = 0;
                            editControl.SelectionLength = editControl.Text.Length;
                        }
                    }
                    return true;
                }
            }
            else if( CurrentCell.ColumnIndex == 5)
            {
                double val=0;
                double.TryParse(string.Concat(CurrentCell.EditedFormattedValue), out val);
                if( val== 0)return base.ProcessDialogKey(keyData);
                ProcessTabKey(keyData);
                if (!this.IsCurrentCellDirty)
                {
                    ProcessTabKey(keyData);
                    ProcessTabKey(keyData);
                    ProcessTabKey(keyData);
                    ProcessTabKey(keyData);                  
                    ProcessF2Key(keyData);
                }
                return true;
            }
            else
            {
                if( key == Keys.Enter)
                {
                    ProcessTabKey(keyData);
                    return true;
                }

             
            }
        }

        return base.ProcessDialogKey(keyData);

    }

    private void subscribeEvents(Control control)
    {
                
        control.KeyDown +=new KeyEventHandler(control_KeyDown);
        control.KeyPress +=new KeyPressEventHandler(control_keyPress);
        control.ControlAdded+=new ControlEventHandler(control_ControlAdded); 
        control.ControlRemoved+=new ControlEventHandler(control_ControlRemoved);
        foreach(Control  innerControl in control.Controls)
        {
            subscribeEvents(innerControl);
        }
    }

    private void unsubscribeEvents(Control control)
    {
         control.KeyDown -=new KeyEventHandler(control_KeyDown);
        control.KeyPress -=new KeyPressEventHandler(control_keyPress);
        control.ControlAdded-=new ControlEventHandler(control_ControlAdded); 
        control.ControlRemoved-=new ControlEventHandler(control_ControlRemoved);
        foreach(Control  innerControl in control.Controls)
        {
            unsubscribeEvents(innerControl);
        }
    }

    private void control_ControlAdded(Object sender ,ControlEventArgs e)
    {
        subscribeEvents(e.Control);
    }

    private void control_ControlRemoved(Object sender, ControlEventArgs e )
    {
        unsubscribeEvents(e.Control);
    }

    private void control_KeyDown(object sender ,KeyEventArgs e )
    {
        keyDownHook(sender, e);
    }
    private void control_keyPress(Object sender ,KeyPressEventArgs e )
    {
        keyPressHook(sender, e);
    }
   

}
}

